File: cliffy-protocols/src/delta.rs
//! State delta computation for efficient synchronization
//!
//! This module provides geometric delta computation between states,
//! enabling bandwidth-efficient synchronization in distributed systems.
//!
//! # Key Concepts
//!
//! - **Delta**: The minimal transformation to go from one state to another
//! - **Compression**: Represent deltas in log space for smaller payloads
//! - **Batching**: Combine multiple deltas into single compound transformations
//!
//! # Example
//!
//! ```rust
//! use cliffy_protocols::delta::{compute_delta, apply_additive_delta};
//! use cliffy_core::GA3;
//!
//! let from = GA3::scalar(1.0);
//! let to = GA3::scalar(5.0);
//!
//! // Compute the delta between states
//! let delta = compute_delta(&from, &to);
//!
//! // Apply delta to reconstruct target state
//! let mut state = from.clone();
//! apply_additive_delta(&mut state, &delta);
//!
//! assert!((state.get(0) - to.get(0)).abs() < 1e-10);
//! ```

use crate::serde_ga3;
use crate::VectorClock;
use cliffy_core::GA3;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// A state delta representing the transformation between two states.
///
/// Deltas can be represented in different forms for efficiency:
/// - `Additive`: Simple difference (to - from)
/// - `Multiplicative`: Versor/rotor transformation
/// - `Compressed`: Log-space representation for smaller payloads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StateDelta {
    /// The delta transformation
    #[serde(with = "serde_ga3")]
    pub transform: GA3,
    /// Type of delta encoding
    pub encoding: DeltaEncoding,
    /// Source state clock (for causal ordering)
    pub from_clock: VectorClock,
    /// Target state clock
    pub to_clock: VectorClock,
    /// Node that computed this delta
    pub source_node: Uuid,
}

/// How the delta is encoded
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum DeltaEncoding {
    /// Simple additive delta: result = state + delta
    Additive,
    /// Multiplicative (sandwich) delta: result = delta * state * reverse(delta)
    Multiplicative,
    /// Compressed log-space: result = exp(delta) * state
    Compressed,
}

impl StateDelta {
    /// Create a new additive delta.
    pub fn additive(
        transform: GA3,
        from_clock: VectorClock,
        to_clock: VectorClock,
        source_node: Uuid,
    ) -> Self {
        Self {
            transform,
            encoding: DeltaEncoding::Additive,
            from_clock,
            to_clock,
            source_node,
        }
    }

    /// Create a new multiplicative (versor) delta.
    pub fn multiplicative(
        transform: GA3,
        from_clock: VectorClock,
        to_clock: VectorClock,
        source_node: Uuid,
    ) -> Self {
        Self {
            transform,
            encoding: DeltaEncoding::Multiplicative,
            from_clock,
            to_clock,
            source_node,
        }
    }

    /// Create a compressed (log-space) delta.
    pub fn compressed(
        transform: GA3,
        from_clock: VectorClock,
        to_clock: VectorClock,
        source_node: Uuid,
    ) -> Self {
        Self {
            transform,
            encoding: DeltaEncoding::Compressed,
            from_clock,
            to_clock,
            source_node,
        }
    }

    /// Get the approximate size of this delta in bytes (for bandwidth estimation).
    pub fn estimated_size(&self) -> usize {
        // 8 coefficients * 8 bytes each + overhead
        8 * 8 + 32
    }

    /// Check if this delta is causally applicable to a state with the given clock.
    pub fn is_applicable_to(&self, state_clock: &VectorClock) -> bool {
        // Delta is applicable if state_clock >= from_clock
        self.from_clock.happens_before(state_clock) || self.from_clock == *state_clock
    }
}

/// Compute the delta between two states.
///
/// Returns an additive delta by default. Use `compute_delta_compressed`
/// for log-space representation.
pub fn compute_delta(from: &GA3, to: &GA3) -> GA3 {
    to - from
}

/// Compute a compressed (log-space) delta.
///
/// This representation is more compact for states that differ by
/// multiplicative factors rather than additive differences.
pub fn compute_delta_compressed(from: &GA3, to: &GA3) -> GA3 {
    // For compressed representation, we compute log(to) - log(from)
    // which can be applied as exp(delta) * from
    //
    // For now, we use a simplified version that works well for
    // scalar-dominated multivectors
    let from_mag = from.magnitude();
    let to_mag = to.magnitude();

    if from_mag < 1e-10 {
        // Can't compute log of zero - fall back to additive
        return to - from;
    }

    // Compute the ratio and take log
    let ratio = to_mag / from_mag;
    let log_ratio = ratio.ln();

    // Return as scalar multivector (simplified)
    GA3::scalar(log_ratio)
}

/// Apply a delta to a state, modifying it in place.
pub fn apply_delta(state: &mut GA3, delta: &StateDelta) {
    match delta.encoding {
        DeltaEncoding::Additive => {
            *state = &*state + &delta.transform;
        }
        DeltaEncoding::Multiplicative => {
            // Sandwich product: delta * state * reverse(delta)
            let rev = delta.transform.reverse();
            *state = delta
                .transform
                .geometric_product(state)
                .geometric_product(&rev);
        }
        DeltaEncoding::Compressed => {
            // Exponential application: exp(delta) * state
            let exp_delta = delta.transform.exp();
            *state = exp_delta.geometric_product(state);
        }
    }
}

/// Apply a raw additive delta to a state.
pub fn apply_additive_delta(state: &mut GA3, delta: &GA3) {
    *state = &*state + delta;
}

/// A batch of deltas that can be applied together.
///
/// Batching reduces network overhead when multiple deltas need to
/// be transmitted.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeltaBatch {
    /// The deltas in this batch, in causal order
    pub deltas: Vec<StateDelta>,
    /// Combined clock covering all deltas
    pub combined_clock: VectorClock,
}

impl DeltaBatch {
    /// Create a new empty batch.
    pub fn new() -> Self {
        Self {
            deltas: Vec::new(),
            combined_clock: VectorClock::new(),
        }
    }

    /// Add a delta to the batch.
    pub fn push(&mut self, delta: StateDelta) {
        self.combined_clock.update(&delta.to_clock);
        self.deltas.push(delta);
    }

    /// Check if the batch is empty.
    pub fn is_empty(&self) -> bool {
        self.deltas.is_empty()
    }

    /// Get the number of deltas in the batch.
    pub fn len(&self) -> usize {
        self.deltas.len()
    }

    /// Combine all additive deltas into a single delta.
    ///
    /// This only works for additive deltas; mixed batches are not combined.
    pub fn combine_additive(&self) -> Option<GA3> {
        if self.deltas.is_empty() {
            return None;
        }

        // Check all deltas are additive
        if !self
            .deltas
            .iter()
            .all(|d| d.encoding == DeltaEncoding::Additive)
        {
            return None;
        }

        // Sum all transforms
        let combined = self
            .deltas
            .iter()
            .fold(GA3::zero(), |acc, d| &acc + &d.transform);

        Some(combined)
    }

    /// Apply all deltas in the batch to a state.
    pub fn apply_to(&self, state: &mut GA3) {
        for delta in &self.deltas {
            apply_delta(state, delta);
        }
    }

    /// Get the estimated total size in bytes.
    pub fn estimated_size(&self) -> usize {
        self.deltas.iter().map(|d| d.estimated_size()).sum()
    }
}

impl Default for DeltaBatch {
    fn default() -> Self {
        Self::new()
    }
}

/// Compute the size savings from using deltas vs full state sync.
///
/// Returns (delta_size, full_size, savings_ratio).
pub fn compute_savings(delta: &StateDelta, _full_state: &GA3) -> (usize, usize, f64) {
    let delta_size = delta.estimated_size();
    let full_size = 8 * 8; // 8 coefficients * 8 bytes

    let savings = if full_size > 0 {
        1.0 - (delta_size as f64 / full_size as f64)
    } else {
        0.0
    };

    (delta_size, full_size, savings)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_additive_delta() {
        let from = GA3::scalar(3.0);
        let to = GA3::scalar(7.0);

        let delta = compute_delta(&from, &to);

        // Delta should be 4.0
        assert!((delta.get(0) - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_apply_additive_delta() {
        let from = GA3::scalar(3.0);
        let to = GA3::scalar(7.0);

        let delta_transform = compute_delta(&from, &to);
        let delta = StateDelta::additive(
            delta_transform,
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        );

        let mut state = from.clone();
        apply_delta(&mut state, &delta);

        assert!((state.get(0) - 7.0).abs() < 1e-10);
    }

    #[test]
    fn test_apply_multiplicative_delta() {
        // Create a rotor-like delta (simplified: scalar 2.0 for doubling)
        let delta = StateDelta::multiplicative(
            GA3::scalar(2.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        );

        let mut state = GA3::scalar(3.0);
        apply_delta(&mut state, &delta);

        // sandwich(2, 3, 2) = 2 * 3 * 2 = 12
        assert!((state.get(0) - 12.0).abs() < 1e-10);
    }

    #[test]
    fn test_apply_compressed_delta() {
        let delta = StateDelta::compressed(
            GA3::scalar(1.0), // e^1 ≈ 2.718
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        );

        let mut state = GA3::scalar(1.0);
        apply_delta(&mut state, &delta);

        // exp(1) * 1 ≈ 2.718
        let expected = std::f64::consts::E;
        assert!((state.get(0) - expected).abs() < 1e-10);
    }

    #[test]
    fn test_delta_batch_combine() {
        let mut batch = DeltaBatch::new();

        batch.push(StateDelta::additive(
            GA3::scalar(1.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        ));

        batch.push(StateDelta::additive(
            GA3::scalar(2.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        ));

        batch.push(StateDelta::additive(
            GA3::scalar(3.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        ));

        let combined = batch.combine_additive().unwrap();

        // 1 + 2 + 3 = 6
        assert!((combined.get(0) - 6.0).abs() < 1e-10);
    }

    #[test]
    fn test_delta_batch_apply() {
        let mut batch = DeltaBatch::new();

        batch.push(StateDelta::additive(
            GA3::scalar(1.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        ));

        batch.push(StateDelta::additive(
            GA3::scalar(2.0),
            VectorClock::new(),
            VectorClock::new(),
            Uuid::new_v4(),
        ));

        let mut state = GA3::scalar(10.0);
        batch.apply_to(&mut state);

        // 10 + 1 + 2 = 13
        assert!((state.get(0) - 13.0).abs() < 1e-10);
    }

    #[test]
    fn test_delta_applicability() {
        let mut from_clock = VectorClock::new();
        let node = Uuid::new_v4();
        from_clock.tick(node);

        let mut to_clock = from_clock.clone();
        to_clock.tick(node);

        let delta = StateDelta::additive(GA3::scalar(1.0), from_clock.clone(), to_clock, node);

        // Delta should be applicable to state with from_clock
        assert!(delta.is_applicable_to(&from_clock));

        // Delta should be applicable to state with higher clock
        let mut higher_clock = from_clock.clone();
        higher_clock.tick(node);
        assert!(delta.is_applicable_to(&higher_clock));
    }

    #[test]
    fn test_compute_delta_compressed() {
        let from = GA3::scalar(2.0);
        let to = GA3::scalar(8.0);

        let delta = compute_delta_compressed(&from, &to);

        // ln(8/2) = ln(4) ≈ 1.386
        let expected = (8.0_f64 / 2.0_f64).ln();
        assert!((delta.get(0) - expected).abs() < 1e-10);
    }

    #[test]
    fn test_delta_encoding_equality() {
        assert_eq!(DeltaEncoding::Additive, DeltaEncoding::Additive);
        assert_ne!(DeltaEncoding::Additive, DeltaEncoding::Multiplicative);
    }
}

File: cliffy-tsukoshi/src/protocols/delta.ts
/**
 * State delta computation for efficient synchronization.
 *
 * Deltas represent minimal transformations between states,
 * enabling bandwidth-efficient sync in distributed systems.
 *
 * @example
 * ```typescript
 * import { computeDelta, applyAdditiveDelta } from 'cliffy-tsukoshi/protocols';
 * import { scalar } from 'cliffy-tsukoshi';
 *
 * const from = scalar(1);
 * const to = scalar(5);
 *
 * // Compute the delta between states
 * const delta = computeDelta(from, to);
 *
 * // Apply delta to reconstruct target state
 * let state = { ...from };
 * state = applyAdditiveDelta(state, delta);
 * ```
 */

import type { GA3 } from '../ga3.js';
import {
  add,
  sub,
  zero,
  scalar,
  magnitude,
  geometricProduct,
  reverse,
  fromCoefficients,
} from '../ga3.js';
import { VectorClock } from './vector-clock.js';

/**
 * How the delta is encoded.
 */
export enum DeltaEncoding {
  /** Simple additive delta: result = state + delta */
  Additive = 'additive',
  /** Multiplicative (sandwich) delta: result = delta * state * reverse(delta) */
  Multiplicative = 'multiplicative',
  /** Compressed log-space: result = exp(delta) * state */
  Compressed = 'compressed',
}

/**
 * A state delta representing the transformation between two states.
 */
export interface StateDelta {
  /** The delta transformation */
  transform: GA3;
  /** Type of delta encoding */
  encoding: DeltaEncoding;
  /** Source state clock (for causal ordering) */
  fromClock: VectorClock;
  /** Target state clock */
  toClock: VectorClock;
  /** Node that computed this delta */
  sourceNode: string;
}

/**
 * Create a new additive delta.
 */
export function additiveDelta(
  transform: GA3,
  fromClock: VectorClock,
  toClock: VectorClock,
  sourceNode: string
): StateDelta {
  return {
    transform,
    encoding: DeltaEncoding.Additive,
    fromClock,
    toClock,
    sourceNode,
  };
}

/**
 * Create a new multiplicative (versor) delta.
 */
export function multiplicativeDelta(
  transform: GA3,
  fromClock: VectorClock,
  toClock: VectorClock,
  sourceNode: string
): StateDelta {
  return {
    transform,
    encoding: DeltaEncoding.Multiplicative,
    fromClock,
    toClock,
    sourceNode,
  };
}

/**
 * Create a compressed (log-space) delta.
 */
export function compressedDelta(
  transform: GA3,
  fromClock: VectorClock,
  toClock: VectorClock,
  sourceNode: string
): StateDelta {
  return {
    transform,
    encoding: DeltaEncoding.Compressed,
    fromClock,
    toClock,
    sourceNode,
  };
}

/**
 * Get the approximate size of a delta in bytes (for bandwidth estimation).
 */
export function estimateDeltaSize(_delta: StateDelta): number {
  // 8 coefficients * 8 bytes each + overhead
  return 8 * 8 + 32;
}

/**
 * Check if a delta is causally applicable to a state with the given clock.
 */
export function isApplicableTo(delta: StateDelta, stateClock: VectorClock): boolean {
  return delta.fromClock.happensBefore(stateClock) || delta.fromClock.equals(stateClock);
}

/**
 * Compute the delta between two states (additive).
 */
export function computeDelta(from: GA3, to: GA3): GA3 {
  return sub(to, from);
}

/**
 * Compute a compressed (log-space) delta.
 *
 * This representation is more compact for states that differ by
 * multiplicative factors rather than additive differences.
 */
export function computeDeltaCompressed(from: GA3, to: GA3): GA3 {
  const fromMag = magnitude(from);
  const toMag = magnitude(to);

  if (fromMag < 1e-10) {
    // Can't compute log of zero - fall back to additive
    return sub(to, from);
  }

  // Compute the ratio and take log
  const ratio = toMag / fromMag;
  const logRatio = Math.log(ratio);

  // Return as scalar multivector (simplified)
  return scalar(logRatio);
}

/**
 * Apply a delta to a state, returning the new state.
 */
export function applyDelta(state: GA3, delta: StateDelta): GA3 {
  switch (delta.encoding) {
    case DeltaEncoding.Additive:
      return add(state, delta.transform);

    case DeltaEncoding.Multiplicative:
      // Sandwich product: delta * state * reverse(delta)
      const rev = reverse(delta.transform);
      const temp = geometricProduct(delta.transform, state);
      return geometricProduct(temp, rev);

    case DeltaEncoding.Compressed:
      // Exponential application: exp(delta) * state
      const expDelta = gaExp(delta.transform);
      return geometricProduct(expDelta, state);

    default:
      throw new Error(`Unknown delta encoding: ${delta.encoding}`);
  }
}

/**
 * Apply a raw additive delta to a state.
 */
export function applyAdditiveDelta(state: GA3, delta: GA3): GA3 {
  return add(state, delta);
}

/**
 * A batch of deltas that can be applied together.
 */
export class DeltaBatch {
  /** The deltas in this batch, in causal order */
  deltas: StateDelta[];
  /** Combined clock covering all deltas */
  combinedClock: VectorClock;

  constructor() {
    this.deltas = [];
    this.combinedClock = new VectorClock();
  }

  /** Add a delta to the batch. */
  push(delta: StateDelta): void {
    this.combinedClock.update(delta.toClock);
    this.deltas.push(delta);
  }

  /** Check if the batch is empty. */
  isEmpty(): boolean {
    return this.deltas.length === 0;
  }

  /** Get the number of deltas in the batch. */
  get length(): number {
    return this.deltas.length;
  }

  /**
   * Combine all additive deltas into a single delta.
   * Returns null if batch contains non-additive deltas.
   */
  combineAdditive(): GA3 | null {
    if (this.deltas.length === 0) {
      return null;
    }

    // Check all deltas are additive
    if (!this.deltas.every(d => d.encoding === DeltaEncoding.Additive)) {
      return null;
    }

    // Sum all transforms
    return this.deltas.reduce(
      (acc, d) => add(acc, d.transform),
      zero()
    );
  }

  /** Apply all deltas in the batch to a state. */
  applyTo(state: GA3): GA3 {
    let result = state;
    for (const delta of this.deltas) {
      result = applyDelta(result, delta);
    }
    return result;
  }

  /** Get the estimated total size in bytes. */
  estimatedSize(): number {
    return this.deltas.reduce((sum, d) => sum + estimateDeltaSize(d), 0);
  }

  /** Serialize to JSON. */
  toJSON(): object {
    return {
      deltas: this.deltas.map(d => ({
        transform: d.transform,
        encoding: d.encoding,
        fromClock: d.fromClock.toJSON(),
        toClock: d.toClock.toJSON(),
        sourceNode: d.sourceNode,
      })),
      combinedClock: this.combinedClock.toJSON(),
    };
  }

  /** Deserialize from JSON. */
  static fromJSON(data: any): DeltaBatch {
    const batch = new DeltaBatch();
    batch.combinedClock = VectorClock.fromJSON(data.combinedClock);
    batch.deltas = data.deltas.map((d: any) => ({
      transform: d.transform,
      encoding: d.encoding,
      fromClock: VectorClock.fromJSON(d.fromClock),
      toClock: VectorClock.fromJSON(d.toClock),
      sourceNode: d.sourceNode,
    }));
    return batch;
  }
}

/**
 * Compute the size savings from using deltas vs full state sync.
 * Returns [deltaSize, fullSize, savingsRatio].
 */
export function computeSavings(delta: StateDelta, _fullState: GA3): [number, number, number] {
  const deltaSize = estimateDeltaSize(delta);
  const fullSize = 8 * 8; // 8 coefficients * 8 bytes

  const savings = fullSize > 0
    ? 1.0 - deltaSize / fullSize
    : 0.0;

  return [deltaSize, fullSize, savings];
}

/**
 * Simple exponential approximation for GA3.
 * Uses Taylor series: exp(x) ≈ 1 + x + x²/2! + x³/3! + ...
 */
function gaExp(mv: GA3): GA3 {
  const coeffs = mv as number[];
  const scalarPart = coeffs[0];

  // For primarily scalar values, use standard exp
  const vectorMag = Math.sqrt(
    coeffs.slice(1).reduce((sum, c) => sum + c * c, 0)
  );

  if (vectorMag < 1e-10) {
    // Pure scalar - standard exponential
    return fromCoefficients([Math.exp(scalarPart), 0, 0, 0, 0, 0, 0, 0]);
  }

  // General case: use Taylor series approximation (4 terms)
  let result = fromCoefficients([1, 0, 0, 0, 0, 0, 0, 0]); // 1
  let term = mv; // x

  result = add(result, term);

  term = scale(geometricProduct(term, mv), 0.5); // x²/2!
  result = add(result, term);

  term = scale(geometricProduct(term, mv), 1 / 3); // x³/3!
  result = add(result, term);

  term = scale(geometricProduct(term, mv), 0.25); // x⁴/4!
  result = add(result, term);

  return result;
}

import { scale } from '../ga3.js';
